File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/C/6886f2fe7b92dcf1b3e045025ca2253f.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_MOD 1000000007
#define MOD 998244353

typedef struct {
    unsigned long long value;
} modint;

modint modint_add(modint a, modint b) {
    modint result;
    result.value = (a.value + b.value) % MAX_MOD;
    return result;
}

modint modint_sub(modint a, modint b) {
    modint result;
    result.value = (a.value - b.value + MAX_MOD) % MAX_MOD;
    return result;
}

modint modint_mul(modint a, modint b) {
    modint result;
    result.value = (a.value * b.value) % MAX_MOD;
    return result;
}

modint modint_pow(modint a, unsigned long long b) {
    modint result;
    result.value = 1;
    while (b > 0) {
        if (b % 2 == 1) {
            result = modint_mul(result, a);
        }
        a = modint_mul(a, a);
        b /= 2;
    }
    return result;
}

typedef struct {
    int first;
    int second;
} pair;

pair make_pair(int a, int b) {
    pair result;
    result.first = a;
    result.second = b;
    return result;
}

pair mp(int a, int b) {
    return make_pair(a, b);
}

int lower_bound(int* arr, int size, int target) {
    int left = 0;
    int right = size;
    while (left < right) {
        int mid = (left + right) / 2;
        if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    return left;
}

int upper_bound(int* arr, int size, int target) {
    int left = 0;
    int right = size;
    while (left < right) {
        int mid = (left + right) / 2;
        if (arr[mid] <= target) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    return left;
}

void solve() {
    int n;
    scanf("%d", &n);
    char t[300001];
    scanf("%s", t);
    int cnter[300001];
    int cnter_size = 0;
    for (int i = 0; i < strlen(t); i++) {
        if (t[i] == '0') {
            cnter[cnter_size++] = i * 2 + i % 2;
        }
    }
    modint hoge[300001];
    modint rev[300001];
    modint nexts;
    nexts.value = 1;
    for (int i = 0; i < cnter_size; i++) {
        hoge[i + 1] = hoge[i];
        hoge[i + 1] = modint_add(hoge[i + 1], modint_mul(nexts, cnter[i] % 2));
        rev[i + 1] = rev[i];
        rev[i + 1] = modint_add(rev[i + 1], modint_mul(nexts, (cnter[i] + 1) % 2));
        nexts = modint_mul(nexts, 2);
    }
    int query;
    scanf("%d", &query);
    for (int i = 0; i < query; i++) {
        int a, b, c;
        scanf("%d %d %d", &a, &b, &c);
        a--; b--;
        pair Left = mp(lower_bound(cnter, cnter_size, a * 2), lower_bound(cnter, cnter_size, (a + c) * 2));
        pair Right = mp(lower_bound(cnter, cnter_size, b * 2), lower_bound(cnter, cnter_size, (b + c) * 2));
        if (Left.second - Left.first != Right.second - Right.first) {
            printf("No\n");
            continue;
        }
        int ok = 1;
        modint aa, bb;
        aa = hoge[Left.second];
        aa = modint_sub(aa, hoge[Left.first]);
        if (a % 2 == b % 2) {
            bb = hoge[Right.second];
            bb = modint_sub(bb, hoge[Right.first]);
        } else {
            bb = rev[Right.second];
            bb = modint_sub(bb, rev[Right.first]);
        }
        if (Left.first != 0) {
            aa = modint_mul(aa, modint_pow(2, Left.first));
        }
        if (Right.first != 0) {
            bb = modint_mul(bb, modint_pow(2, Right.first));
        }
        if (aa.value != bb.value) {
            ok = 0;
        }
        if (ok == 1) {
            printf("Yes\n");
        } else {
            printf("No\n");
        }
    }
}

int main() {
    solve();
    return 0;
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Go/6886f2fe7b92dcf1b3e045025ca2253f.go
package main

import (
	"fmt"
	"math/big"
)

const (
	MAX_MOD = 1000000007
	MOD     = 998244353
)

type modint struct {
	value *big.Int
	mod   *big.Int
}

func newModint(a int64, mod int64) *modint {
	return &modint{
		value: big.NewInt(a),
		mod:   big.NewInt(mod),
	}
}

func (m *modint) add(rhs *modint) *modint {
	ans := new(big.Int).Add(m.value, rhs.value)
	ans.Mod(ans, m.mod)
	return &modint{
		value: ans,
		mod:   m.mod,
	}
}

func (m *modint) sub(rhs *modint) *modint {
	ans := new(big.Int).Sub(m.value, rhs.value)
	ans.Mod(ans, m.mod)
	return &modint{
		value: ans,
		mod:   m.mod,
	}
}

func (m *modint) mul(rhs *modint) *modint {
	ans := new(big.Int).Mul(m.value, rhs.value)
	ans.Mod(ans, m.mod)
	return &modint{
		value: ans,
		mod:   m.mod,
	}
}

func (m *modint) div(rhs *modint) *modint {
	inv := new(big.Int).ModInverse(rhs.value, m.mod)
	ans := new(big.Int).Mul(m.value, inv)
	ans.Mod(ans, m.mod)
	return &modint{
		value: ans,
		mod:   m.mod,
	}
}

func (m *modint) pow(b int64) *modint {
	ans := new(big.Int).Exp(m.value, big.NewInt(b), m.mod)
	return &modint{
		value: ans,
		mod:   m.mod,
	}
}

func (m *modint) equal(rhs *modint) bool {
	return m.value.Cmp(rhs.value) == 0
}

func (m *modint) less(rhs *modint) bool {
	return m.value.Cmp(rhs.value) < 0
}

func (m *modint) String() string {
	return m.value.String()
}

type pair struct {
	first  *modint
	second *modint
}

func newPair(a, b *modint) *pair {
	return &pair{
		first:  a,
		second: b,
	}
}

func solve() {
	var n int
	fmt.Scan(&n)
	var t string
	fmt.Scan(&t)

	cnter := make([]int, 0)
	for i := 0; i < len(t); i++ {
		if t[i] == '0' {
			cnter = append(cnter, i*2+i%2)
		}
	}

	hoge := make([]*pair, 300000)
	rev := make([]*pair, 300000)
	nexts := newPair(newModint(1, MAX_MOD), newModint(1, MOD))
	for i := 0; i < len(cnter); i++ {
		hoge[i+1] = newPair(hoge[i].first.add(nexts.first.mul(newModint(int64(cnter[i]%2), MAX_MOD))), hoge[i].second.add(nexts.second.mul(newModint(int64(cnter[i]%2), MOD))))
		rev[i+1] = newPair(rev[i].first.add(nexts.first.mul(newModint(int64((cnter[i]+1)%2), MAX_MOD))), rev[i].second.add(nexts.second.mul(newModint(int64((cnter[i]+1)%2), MOD))))
		nexts.first = nexts.first.mul(newModint(2, MAX_MOD))
		nexts.second = nexts.second.mul(newModint(2, MOD))
	}

	var query int
	fmt.Scan(&query)
	for i := 0; i < query; i++ {
		var a, b, c int
		fmt.Scan(&a, &b, &c)
		a--
		b--
		left := []int{lowerBound(cnter, a*2), lowerBound(cnter, (a+c)*2)}
		right := []int{lowerBound(cnter, b*2), lowerBound(cnter, (b+c)*2)}
		if left[1]-left[0] != right[1]-right[0] {
			fmt.Println("No")
			continue
		}
		ok := 1
		aa := newPair(hoge[left[1]].first.sub(hoge[left[0]].first), hoge[left[1]].second.sub(hoge[left[0]].second))
		if a%2 == b%2 {
			bb := newPair(hoge[right[1]].first.sub(hoge[right[0]].first), hoge[right[1]].second.sub(hoge[right[0]].second))
			aa.first = aa.first.div(hoge[left[0]].first.pow(left[0]))
			aa.second = aa.second.div(hoge[left[0]].second.pow(left[0]))
			bb.first = bb.first.div(hoge[right[0]].first.pow(right[0]))
			bb.second = bb.second.div(hoge[right[0]].second.pow(right[0]))
			if !aa.first.equal(bb.first) || !aa.second.equal(bb.second) {
				ok = 0
			}
		} else {
			bb := newPair(rev[right[1]].first.sub(rev[right[0]].first), rev[right[1]].second.sub(rev[right[0]].second))
			aa.first = aa.first.div(hoge[left[0]].first.pow(left[0]))
			aa.second = aa.second.div(hoge[left[0]].second.pow(left[0]))
			bb.first = bb.first.div(rev[right[0]].first.pow(right[0]))
			bb.second = bb.second.div(rev[right[0]].second.pow(right[0]))
			if !aa.first.equal(bb.first) || !aa.second.equal(bb.second) {
				ok = 0
			}
		}
		if ok == 1 {
			fmt.Println("Yes")
		} else {
			fmt.Println("No")
		}
	}
}

func lowerBound(arr []int, target int) int {
	left := 0
	right := len(arr)
	for left < right {
		mid := (left + right) / 2
		if arr[mid] < target {
			left = mid + 1
		} else {
			right = mid
		}
	}
	return left
}

func main() {
	solve()
}
